File: Try_GameEngine5/src/com/example/try_gameengine/utils/GameTimeUtil.java
package com.example.try_gameengine.utils;

public class GameTimeUtil {
	private long initTimeMS;
	private long intervalMS;
	private long nextExecuteTimeMS;
	private boolean enable = true;
	
	public GameTimeUtil(long intervalMS){
		this.initTimeMS = System.currentTimeMillis();
		this.intervalMS = intervalMS;
		this.nextExecuteTimeMS = initTimeMS + intervalMS;
	}
	
	public GameTimeUtil(long initTimeMS, long intervalMS){
		this.initTimeMS = initTimeMS;
		this.intervalMS = intervalMS;
		this.nextExecuteTimeMS = initTimeMS + intervalMS;
	}
	
	public boolean isArriveExecuteTime(){
		if(!enable)
			return false;
		boolean isArriveTime = false;
		if(System.currentTimeMillis() >= nextExecuteTimeMS){
			isArriveTime = true;
			nextExecuteTimeMS += intervalMS;
		}
		return isArriveTime;
	}
	
	public boolean isArriveExecuteTime(long currentTimeMS){
		if(!enable)
			return false;
		boolean isArriveTime = false;
		if(currentTimeMS >= nextExecuteTimeMS){
			isArriveTime = true;
			nextExecuteTimeMS += intervalMS;
		}
		return isArriveTime;
	}
	
	public boolean isArriveExecuteTimeIfOneDelayThenAllDelay(){
		return isArriveExecuteTimeIfOneDelayThenAllDelay(System.currentTimeMillis());
	}
	
	public boolean isArriveExecuteTimeIfOneDelayThenAllDelay(long currentTimeMS){
		if(!enable)
			return false;
		boolean isArriveTime = false;
		if(currentTimeMS >= nextExecuteTimeMS){
			isArriveTime = true;
			nextExecuteTimeMS = currentTimeMS + intervalMS;
		}
		return isArriveTime;
	}
	
	public void enable(boolean enable){
		this.enable = enable;
	}
}

File: Try_GameEngine5/src/com/example/try_gameengine/utils/GameTimeUtil.kt
package com.example.try_gameengine.utils

class GameTimeUtil {
    private val initTimeMS: Long
    private val intervalMS: Long
    private var nextExecuteTimeMS: Long
    private var enable = true

    constructor(intervalMS: Long) {
        this.initTimeMS = System.currentTimeMillis()
        this.intervalMS = intervalMS
        this.nextExecuteTimeMS = initTimeMS + intervalMS
    }

    constructor(initTimeMS: Long, intervalMS: Long) {
        this.initTimeMS = initTimeMS
        this.intervalMS = intervalMS
        this.nextExecuteTimeMS = initTimeMS + intervalMS
    }

    val isArriveExecuteTime: Boolean
        get() {
            if (!enable) return false
            var isArriveTime = false
            if (System.currentTimeMillis() >= nextExecuteTimeMS) {
                isArriveTime = true
                nextExecuteTimeMS += intervalMS
            }
            return isArriveTime
        }

    fun isArriveExecuteTime(currentTimeMS: Long): Boolean {
        if (!enable) return false
        var isArriveTime = false
        if (currentTimeMS >= nextExecuteTimeMS) {
            isArriveTime = true
            nextExecuteTimeMS += intervalMS
        }
        return isArriveTime
    }

    val isArriveExecuteTimeIfOneDelayThenAllDelay: Boolean
        get() = isArriveExecuteTimeIfOneDelayThenAllDelay(System.currentTimeMillis())

    fun isArriveExecuteTimeIfOneDelayThenAllDelay(currentTimeMS: Long): Boolean {
        if (!enable) return false
        var isArriveTime = false
        if (currentTimeMS >= nextExecuteTimeMS) {
            isArriveTime = true
            nextExecuteTimeMS = currentTimeMS + intervalMS
        }
        return isArriveTime
    }

    fun enable(enable: Boolean) {
        this.enable = enable
    }
}
